File: python/CountSemiprimes.py
#https://codility.com/demo/results/training56JA7K-E86/
# you can write to stdout for debugging purposes, e.g.
# print "this is a debug message"
def factorization(N):
    f = [0] * (N + 1)
    f[0] = f[1] = -1
    i = 2
    while i * i <= N:
        if f[i] == 0:
            k = i * i
            while k <= N:
                if f[k] == 0:
                   f[k] = i
                k += i
        i += 1
    return f
def solution(N, P, Q):
    # write your code in Python 2.7
    ans = []
    prefixSum = [0] * (N + 1)
    fact = factorization(N)
    for i in xrange(2, N+1):
        if fact[i] != 0 and fact[i // fact[i]] == 0:
            prefixSum[i] = prefixSum[i - 1] + 1
        else:
            prefixSum[i] = prefixSum[i - 1]
    for i in xrange(len(P)):
        ans += [prefixSum[Q[i]] - prefixSum[P[i] - 1]]
    return ans

File: java/CountSemiprimes.java
//https://codility.com/demo/results/trainingPC92RP-4FE/
class SolutionCountSemiprimes {
    private int[] factorization(int N) {
        int[] fact = new int[N + 1];
        fact[0] = fact[1] = 1;
        int i = 2, k = 0;
        while(i * i <= N) {
            if(fact[i] == 0){
                k = i * i;
                while(k <= N) {
                    if(fact[k] == 0) {
                        fact[k] = i;
                    }
                    k += i;
                }
            }
            ++i;
        }
        return fact;
    }
    public int[] solution(int N, int[] P, int[] Q) {
        // write your code in Java SE 8
        int[] factors = factorization(N);
        int[] memo = new int[N + 1];
        int len = P.length;
        int[] ans = new int[len];
        memo[0] = memo[1] = 0;
        for(int i = 2; i < N + 1; ++i) {
            memo[i] = memo[i - 1];
            if(factors[i] != 0 && factors[i / factors[i]] == 0)
                ++memo[i];
        }
        for(int i = 0; i< len; ++i)
            ans[i] = memo[Q[i]] - memo[P[i] - 1];
        return ans;
    }
}
